src/libraries/System.Private.CoreLib/src/System/Reflection/Emit/TypeNameBuilder.cs

```diff
@@ -0,0 +1,429 @@
+
+        internal enum Format
+        {
+            ToString,
+            FullName,
+            AssemblyQualifiedName,
+        }
+
+        internal static string? ToString(Type type, Format format)
+        {
+            if (format == Format.FullName || format == Format.AssemblyQualifiedName)
+            {
+                if (!type.IsGenericTypeDefinition && type.ContainsGenericParameters)
+                    return null;
+            }
+
+            TypeNameBuilder tnb = new TypeNameBuilder();
+            ConstructAssemblyQualifiedNameWorker(tnb, type, format);
+            return tnb.ToString();
+        }
+
+        private static void AddElementType(TypeNameBuilder tnb, Type elementType)
+        {
+            if (elementType.HasElementType)
+                AddElementType(tnb, elementType.GetElementType()!);
+
+            if (elementType.IsPointer)
+                tnb.AddPointer();
+            else if (elementType.IsByRef)
+                tnb.AddByRef();
+            else if (elementType.IsSZArray)
+                tnb.AddSzArray();
+            else if (elementType.IsArray)
+                tnb.AddArray(elementType.GetArrayRank());
+        }
+
+        private static void ConstructAssemblyQualifiedNameWorker(TypeNameBuilder tnb, Type type, Format format)
+        {
+            Type rootType = type;
+
+            while (rootType.HasElementType)
+                rootType = rootType.GetElementType()!;
+
+            // Append namespace + nesting + name
+            List<Type> nestings = new List<Type>();
+            for (Type? t = rootType; t != null; t = t.IsGenericParameter ? null : t.DeclaringType)
+                nestings.Add(t);
+
+            for (int i = nestings.Count - 1; i >= 0; i--)
+            {
+                Type enclosingType = nestings[i];
+                string name = enclosingType.Name;
+
+                if (i == nestings.Count - 1 && enclosingType.Namespace != null && enclosingType.Namespace.Length != 0)
+                    name = enclosingType.Namespace + "." + name;
+
+                tnb.AddName(name);
+            }
+
+            // Append generic arguments
+            if (rootType.IsGenericType && (!rootType.IsGenericTypeDefinition || format == Format.ToString))
+            {
+                Type[] genericArguments = rootType.GetGenericArguments();
+
+                tnb.OpenGenericArguments();
+                for (int i = 0; i < genericArguments.Length; i++)
+                {
+                    Format genericArgumentsFormat = format == Format.FullName ? Format.AssemblyQualifiedName : format;
+
+                    tnb.OpenGenericArgument();
+                    ConstructAssemblyQualifiedNameWorker(tnb, genericArguments[i], genericArgumentsFormat);
+                    tnb.CloseGenericArgument();
+                }
+                tnb.CloseGenericArguments();
+            }
+
+            // Append pointer, byRef and array qualifiers
+            AddElementType(tnb, type);
```

Comment:
It looks odd that `AddElementType` is called even for types without element type. I think it would look better like this:
```
if (type.HasElementType)
    AddElementType(type);
```

Reply:
Your suggestion seems like it would change behavior - even if type.HasElementType is false, AddElementType can emit the pointer/ref/array symbols for type itself.